src/opt/ReplaceHalfMaskedMemOps.cpp

```diff
@@ -0,0 +1,216 @@
+//  %ptr = bitcast %v8_uniform_FVector4f* %Result.i to <4 x float>*
+//  store <4 x float> %vec.half, <4 x float>* %ptr
+//
+void lReplaceMaskedStore(llvm::IRBuilder<> &B, llvm::CallInst *CI) {
+    llvm::Value *vec = CI->getOperand(0);
+    llvm::Value *ptr = CI->getOperand(1);
+    int alignment = llvm::dyn_cast<llvm::ConstantInt>(CI->getOperand(2))->getZExtValue();
+
+    B.SetInsertPoint(CI);
+
+    vec = lShrinkVector(B, vec);
+    ptr = lBitcastPointerType(B, ptr, vec);
+    llvm::Value *store = B.CreateAlignedStore(vec, ptr, llvm::Align(alignment));
+
+    LLVMCopyMetadata(store, CI);
+    CI->eraseFromParent();
+}
+
+// This function replaces, e.g.,
+//
+// %ptr = bitcast float* %A to <8 x float>*
+// %vec = call <8 x float> @llvm.masked.load.v8f32.p0(ptr %ptr, i32 1,
+//      <8 x i1> <i1 true, i1 true, i1 true, i1 true, i1 false, i1 false, i1 false, i1 false>,
+//      <8 x float> <float poison, float poison, float poison, float poison, float 0, float 0, float 0, float 0>)
+//
+// with
+//
+// %ptr = bitcast float* %A to <4 x float>*
+// %vec.half = load <4 x float>, <4 x float>* %ptr
+// %vec = shufflevector <4 x float> %vec.half, <4 x float> <float 0, float 0, float 0, float 0>,
+//                      <8 x i32> <i32 0, i32 1, i32 2, i32 3, i32 4, i32 5, i32 6, i32 7>
+//
+void lReplaceMaskedLoad(llvm::IRBuilder<> &B, llvm::CallInst *CI) {
+    llvm::Function *F = CI->getCalledFunction();
+    llvm::Value *ptr = CI->getOperand(0);
+    llvm::Value *passthrough = CI->getOperand(3);
+
+    B.SetInsertPoint(CI);
+
+    passthrough = lShrinkConstVec(F->getParent()->getContext(), passthrough);
+    ptr = lBitcastPointerType(B, ptr, passthrough);
+
+    auto *vecType = passthrough->getType();
+    llvm::Twine origName = CI->getName();
+    llvm::Value *half = B.CreateLoad(vecType, ptr, origName + ".half");
+
+    llvm::Value *replacement = lMergeVectors(B, half, passthrough, origName);
+
+    LLVMCopyMetadata(replacement, CI);
+    CI->replaceAllUsesWith(replacement);
+    CI->eraseFromParent();
+}
+
+llvm::PreservedAnalyses ReplaceHalfMaskedMemOpsPass::run(llvm::Function &F, llvm::FunctionAnalysisManager &FAM) {
+    if (F.empty()) {
+        return llvm::PreservedAnalyses::all();
+    }
+
+    llvm::IRBuilder<> builder(F.getParent()->getContext());
+    std::vector<llvm::CallInst *> storesToReplace;
+    std::vector<llvm::CallInst *> loadsToReplace;
+    for (auto &BB : F) {
+        for (auto &I : BB) {
+            auto *CI = llvm::dyn_cast<llvm::CallInst>(&I);
+            if (!CI) {
+                continue;
+            }
+
+            llvm::Function *CF = CI->getCalledFunction();
+            if (!(CF && CF->isIntrinsic())) {
+                continue;
+            }
+
+            if (CF->getIntrinsicID() == llvm::Intrinsic::masked_store) {
+                llvm::Value *mask = CI->getOperand(3);
+                if (lIsSecondHalfAllFalse(mask)) {
```

Comment:
Here and in other similar places, please check that `mask != nullptr`.
I suggest running static analysis on this PR before merging.

Reply:
I am not quite sure why it is needed. Doesn't sane LLVM IR guarantee that mask is non-null? Moreover, nullptr is safe here because inside the next function call we just return `false` for it.